### PR TITLE
Add magic method to our TF models to convert datasets with column inference

### PR DESCRIPTION
**Pull request by Rocketknight1**
Left to do:

- [x] Add docstring
- [x] Figure out what type to set for `dataset` (since `datasets` isn't imported)
- [x] Set default values for `batch_size` / `shuffle`?
- [x] Do we need to document this anywhere besides the docstring?
- [x] Anything else I forgot? (Reviewers please yell at me)

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hey all, this is the method in `transformers` that I moved all the column inference code to! I also need some advice:

- How do we handle an input type of `Dataset` when `datasets` may not be imported?
- Should this be moved to a utility function that takes `model` as an argument rather than a method on `TFPreTrainedModel`?
- Should I set default values for any arguments to reduce the amount of typing users have to do?
- How should I make sure users find out about this (besides revamping the examples and notebooks once it's merged)?

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
...
+    ):
+        """
+        Wraps a HuggingFace `Dataset` as a `tf.data.Dataset` with collation and batching. This method is designed to
+        create a "ready-to-use" dataset that can be passed directly to Keras methods like `fit()` without further
+        modification. The method will drop columns from the dataset if they don't match input names for the model. If
+        you want to specify the column names to return rather than using the names that match this model, we recommend
+        using `Dataset.to_tf_dataset()` instead.
+
+        Args:
+            dataset (`Any`):
```
Comment (gante) [inline src/transformers/modeling_tf_utils.py:907]:
```suggestion
            dataset (`ArrowDataset`):
```
(since the docstring doesn't need the import)

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
         # set it directly, but the user can pass it to fit().
         return {"epoch": extra_data["epoch"]}
 
+    def prepare_tf_dataset(
+        self,
+        dataset: Any,
```
Comment (gante) [inline src/transformers/modeling_tf_utils.py:890]: Since this function is to be used in tandem with `datasets`, perhaps this function could be entirely import-protected by `is_datasets_available`. Then we could put the right type here.

(Alternatively, you can use a [forward reference](https://peps.python.org/pep-0484/#forward-references) here. The information would be nice, but I don't know if our docs play well with that 🤔)

File: tests/test_modeling_tf_common.py [hunk 3]
```diff
@@ -1479,6 +1481,59 @@ def test_model_main_input_name(self):
...
+                # Assert we discarded the unwanted extra column but kept everything else
+                self.assertEqual(feature_columns + label_columns, len(input_dataset.features) - 1)
+                if isinstance(test_batch, dict):
+                    self.assertNotIn("extra_unwanted_column", test_batch)
+                if isinstance(test_batch_labels, dict):
+                    self.assertNotIn("extra_unwanted_column", test_batch_labels)
+                model.compile(optimizer="sgd", run_eagerly=True)
+                if isinstance(test_batch_labels, dict) and "aggregation_labels" in test_batch_labels:
+                    breakpoint()
+                    print()
```
Comment (gante) [inline tests/test_modeling_tf_common.py:1534]: Probably unwanted :D

**Review by gante — COMMENTED** [review]
In general looks good -- the biggest question to settle is the import requirement of `datasets`

**Comment by gante** [comment]
> Should this be moved to a utility function that takes model as an argument rather than a method on TFPreTrainedModel?

All our models inherit from `TFPreTrainedModel`, so it should be fine

> Should I set default values for any arguments to reduce the amount of typing users have to do?

I'd keep the same defaults as in `datasets`, for a consistent experience. But I'm not the most experienced in this domain :D

> How should I make sure users find out about this (besides revamping the examples and notebooks once it's merged)?

We are getting to the point where we have a lot of content to announce (these changes, the metrics working correctly, generate and its updates, new models, ...), maybe we can start a once-a-week TF communication of some sort!

File: src/transformers/modeling_tf_utils.py [hunk 4]
```diff
@@ -35,6 +35,7 @@
 from huggingface_hub import Repository, list_repo_files
 from requests import HTTPError
 
+from . import DataCollatorWithPadding, DefaultDataCollator, PreTrainedTokenizer
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:38]: Not sure why there is the dependency on tokenization utils but if it's just for type hints, it should be protected by an `if TYPE_CHECKING`

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/modeling_tf_utils.py:890]: Yes, please use a forward reference here.

File: src/transformers/modeling_tf_utils.py [hunk 5]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
         # set it directly, but the user can pass it to fit().
         return {"epoch": extra_data["epoch"]}
 
+    def prepare_tf_dataset(
+        self,
+        dataset: Any,
+        batch_size: int,
+        shuffle: bool,
+        tokenizer: Optional[PreTrainedTokenizer] = None,
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:893]:
```suggestion
        tokenizer: Optional["PreTrainedTokenizerBase"] = None,
```

File: src/transformers/modeling_tf_utils.py [hunk 6]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
...
+        collate_fn: Optional[Callable] = None,
+        collate_fn_args: Optional[dict] = None,
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:895]:
```suggestion
        collate_fn_args: Optional[Dict[str, Any]] = None,
```

File: src/transformers/modeling_tf_utils.py [hunk 7]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
...
+        drop_remainder: Optional[bool] = None,
+        prefetch: bool = True,
+    ):
+        """
+        Wraps a HuggingFace `Dataset` as a `tf.data.Dataset` with collation and batching. This method is designed to
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:900]:
```suggestion
        Wraps a HuggingFace `datasets.Dataset` as a `tf.data.Dataset` with collation and batching. This method is designed to
```
and also add a link to the doc

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_tf_utils.py:907]: Please use the full path (datasets.ArrowDataset ?)

File: src/transformers/modeling_tf_utils.py [hunk 8]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
...
+                Whether to return samples from the dataset in random order. Usually `True` for training datasets and
+                `False` for validation/test datasets.
+            tokenizer (`PreTrainedTokenizer`, *optional*):
+                A `PreTrainedTokenizer` that will be used to pad samples to create batches. Has no effect if a specific
+                `collate_fn` is passed instead.
+            collate_fn (`Callable`, *optional*):
+                A function that collates samples from the dataset into a single batch. Defaults to
+                `DefaultDataCollator` if no `tokenizer` is supplied or `DataCollatorWithPadding` if a `tokenizer` is
+                passed.
+            collate_fn_args (`dict`, *optional*):
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:921]:
```suggestion
            collate_fn_args (`Dict[str, Any]`, *optional*):
```

**Review by sgugger — COMMENTED** [review]
Thanks for working on this! Let's polish the type annotations/imports and it should be good to go.

File: src/transformers/dependency_versions_table.py [hunk 9]
```diff
@@ -44,7 +44,7 @@
     "pytest": "pytest",
     "pytest-timeout": "pytest-timeout",
     "pytest-xdist": "pytest-xdist",
-    "python": "python>=3.6.0",
+    "python": "python>=3.7.0",
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/dependency_versions_table.py:47]: Yep, let me try to make it go away!

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:907]: Fixed!

File: tests/test_modeling_tf_common.py (hunk 3 above)
Comment (Rocketknight1) [reply tests/test_modeling_tf_common.py]: Fixed >.>

File: src/transformers/modeling_tf_utils.py [hunk 10]
```diff
@@ -884,6 +885,93 @@ def load_repo_checkpoint(self, repo_path_or_name):
...
+            if tokenizer is None:
+                collate_fn = DefaultDataCollator(return_tensors="tf")
+            else:
+                collate_fn = DataCollatorWithPadding(tokenizer=tokenizer, return_tensors="tf")
+        if collate_fn_args is None:
+            collate_fn_args = dict()
+        try:
+            from datasets import Dataset
+        except ImportError:
+            raise ImportError("Called a datasets-specific method but datasets cannot be imported!")
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Done!

File: src/transformers/dependency_versions_table.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/dependency_versions_table.py]: Fixed by rebasing

File: src/transformers/modeling_tf_utils.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:38]: Done!

File: src/transformers/modeling_tf_utils.py [hunk 11]
```diff
@@ -892,6 +898,94 @@ def load_repo_checkpoint(self, repo_path_or_name):
...
+        further modification. The method will drop columns from the dataset if they don't match input names for the
+        model. If you want to specify the column names to return rather than using the names that match this model, we
+        recommend using `Dataset.to_tf_dataset()` instead.
+
+        Args:
+            dataset (`Any`):
+                A `datasets.Dataset` to be wrapped as a `tf.data.Dataset`.
+            batch_size (`int`, defaults to *8*):
+                The size of batches to return.
+            shuffle (`bool`, defaults to *True*):
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:924]:
```suggestion
            batch_size (`int`, defaults to 8):
                The size of batches to return.
            shuffle (`bool`, defaults to `True`):
```
Values that are numbers are left as is. Other values are in code blocks, not italics.

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py]: Got it, sorry!

**Review by gante — APPROVED** [review]

**Review by sgugger — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
Looks quite cool!

**Comment by Rocketknight1** [comment]
Quick update: I think this is ready to merge, but I've only really tested it with the updated `to_tf_dataset()` method in datasets, which hasn't been merged yet (but is due very soon!). As such, I don't want to merge it until that's in, because there could be edge case issues with the old method that I haven't seen.
